### PR TITLE
KOGITO-2741 Enrich TraceEvent with data types information

### PR DESCRIPTION
**Pull request by kostola**
- [Related JIRA ticket](https://issues.redhat.com/browse/KOGITO-2741)
- [Related PR in kogito-apps](https://github.com/kiegroup/kogito-apps/pull/345)

### CODE REVIEW

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java [hunk 1]
```diff
@@ -313,51 +327,64 @@ private static long computeDurationMillis(EvaluateEvent beginEvent, EvaluateEven
...
-                .map(EventUtils::traceTypeFrom)
                 .orElse(null);
 
-        JsonNode value = Optional.ofNullable(decisionResult.getResult())
-                .<JsonNode>map(Json.MAPPER::valueToTree)
-                .orElse(null);
+        // cast to DMNBaseNode here is required to have access to getDependencies method
+        Map<String, DMNType> decisionInputTypes = Optional.ofNullable(model)
+                .map(m -> m.getDecisionById(decisionResult.getDecisionId()))
+                .filter(DMNBaseNode.class::isInstance)
```
Comment (danielezonca) [inline addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java:352]: Can you please create a ticket to review this code (it should be nice to use only interfaces)?

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java [hunk 2]
```diff
@@ -138,10 +141,24 @@ private static URI buildSource(EvaluateEvent event) {
...
+
+        return Stream.concat(
+                firstEvent.getContext().entrySet().stream()
+                        .map(entry -> inputNodesMap.containsKey(entry.getKey())
+                                ? traceInputFrom(inputNodesMap.get(entry.getKey()), entry.getValue())
+                                : traceInputFrom(entry.getKey(), entry.getValue())
+                        ),
+                inputNodesMap.entrySet().stream()
+                        .filter(entry -> !firstEvent.getContext().containsKey(entry.getKey()))
+                        .map(entry -> traceInputFrom(entry.getValue(), null))
```
Comment (danielezonca) [inline addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java:155]: Can you please split this code in multiple methods to improve readability?

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java (hunk 1 above)
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java:352]: Ticket created ([link](https://issues.redhat.com/browse/KOGITO-2874))

File: addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java (hunk 2 above)
Comment (kostola) [reply addons/tracing/tracing-decision-common/src/main/java/org/kie/kogito/tracing/decision/aggregator/DefaultAggregator.java]: Done :+1:

**Comment by danielezonca** [comment]
jenkins retest this please

**Comment by danielezonca** [comment]
jenkins retest this please

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.kie.kogito.tracing.decision.event.variable;
+
+import java.util.Collection;
+
+public class CollectionVariable extends TypedVariable {
```
Comment (r00ta) [inline addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java:21]: Could you add a complex test that is using collections and more complex scenarios like collections of structures?

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/StructureVariable.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ *  limitations under the License.
+ */
+
+package org.kie.kogito.tracing.decision.event.variable;
+
+import java.util.Map;
+
+public class StructureVariable extends TypedVariable {
+
+    private Map<String, TypedVariable> value;
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/StructureVariable.java:23]: This is actually needed: it reflects the API defined in the PoC, which is one of the goals of this task.

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    private Collection<TypedVariable> value;
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java:23]: [Same as here](https://github.com/kiegroup/kogito-runtimes/pull/651#discussion_r471486549)

**Comment by danielezonca** [comment]
jenkins retest this please

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/TypedVariable.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        defaultImpl = TypedVariable.Kind.class,
+        property = "kind",
+        visible = true
+)
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = UnitVariable.class, name = "UNIT"),
+        @JsonSubTypes.Type(value = CollectionVariable.class, name = "COLLECTION"),
+        @JsonSubTypes.Type(value = StructureVariable.class, name = "STRUCTURE")
+})
+public abstract class TypedVariable {
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/TypedVariable.java:36]: Done on all subclasses

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/StructureVariable.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.kie.kogito.tracing.decision.event.variable;
+
+import java.util.Map;
+
+public class StructureVariable extends TypedVariable {
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/StructureVariable.java:25]: Done

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java:25]: Done

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/UnitVariable.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.kie.kogito.tracing.decision.event.variable;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.JsonNode;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
+
+public class UnitVariable extends TypedVariable {
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/UnitVariable.java:27]: Done

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/UnitVariable.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    @JsonInclude(NON_NULL)
+    private String baseType;
+
+    private JsonNode value;
```
[the remark this answers is not in the document]
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/UnitVariable.java:34]: Done

**Comment by danielezonca** [comment]
jenkins retest this please

File: addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java (hunk 3 above)
Comment (kostola) [reply addons/tracing/tracing-decision-api/src/main/java/org/kie/kogito/tracing/decision/event/variable/CollectionVariable.java:25]: I created a [dedicated task](https://issues.redhat.com/browse/KOGITO-3140) for this.

**Pull request merged**